### PR TITLE
Slideshow

### PR DESCRIPTION
**Pull request by piuccio**
Display slideshow of images on a front.

Slideshow must be manually curated by editors, they can choose up to 5 images to display.

Everything is behind a switch and an A/B test

### CODE REVIEW

**Comment by paperboyo** [comment]
OMGOMG!

Could kindly I ask for a screenshot? Could ‘article’ above mean a gallery (rarely normal articles have 5 good images)? How/where are they ‘manually curated’? Will be anxiously awaiting this on CODE...

Regards
Mateusz

**Comment by piuccio** [comment]
`article` means anything, not only galleries. It works like a replace image in fronts tool, but instead of a single image you can select 5.

ezgif com-optimize

**Comment by paperboyo** [comment]
Thanks! Will this be on dev version of fronts tool, before it makes it's way to PROD? (I just wanna have a play.) (Ok, ok, and moan a bit, too ;-)

**Comment by piuccio** [comment]
Yes, the plan is to put it on CODE first, it's also behind an A/B test. We'll gather some feedback and then turn it on on PROD. I'll let you know when deployed.

**Comment by paperboyo** [comment]
Superb! Thank you kindly!

File: common/app/model/content.scala [hunk 1]
```diff
@@ -262,6 +265,24 @@ class Content protected (val apiContent: ApiContentWithMeta) extends Trail with
...
+      w = for {
+        a <- z.asInstanceOf[JsArray].value
+        y = for {
+          src <- a.asInstanceOf[JsObject].\("src").asOpt[String]
+          width <- a.asInstanceOf[JsObject].\("width").asOpt[String].flatMap(s => Try(s.toInt).toOption)
+          height <- a.asInstanceOf[JsObject].\("height").asOpt[String].flatMap(s => Try(s.toInt).toOption)
+        } yield FaciaImageElement(src, width, height)
+      } yield y
+    } yield w).getOrElse(Nil).flatten
+
```
Comment (JustinPinner) [inline common/app/model/content.scala]: Can we use more meaningful names? x, y, z etc save space but it's not immediately clear what we're yielding.

File: common/app/model/trails.scala [hunk 2]
```diff
@@ -45,9 +47,16 @@ trait Trail extends Elements with Tags with FaciaFields with Dates {
     case snap: Snap => snap.snapHref.filter(_.nonEmpty)
     case t: Trail => Option(t.url)
   }
+
+  def trailSlideshow(aspectWidth: Int, aspectHeight: Int): Iterable[FaciaImageElement] =
+  slideshow
+    .filter(image => IsRatio(5, 3, image.width, image.height))
```
Comment (JustinPinner) [inline common/app/model/trails.scala]: Maybe I'm missing something, but should 5, 3 actually be aspectWidth, aspectHeight?

File: common/app/model/content.scala (hunk 1 above)
Comment (johnduffell) [inline common/app/model/content.scala]: good point, will improve this

File: facia-press/app/frontpress/jsonModels.scala [hunk 3]
```diff
@@ -108,10 +110,11 @@ case class ItemMeta(
   snapUri:       Option[JsValue],
   showKickerTag: Option[JsValue],
   showKickerSection: Option[JsValue],
-  showMainVideo: Option[JsValue]
+  showMainVideo: Option[JsValue],
+  slideshowImages: Option[Boolean],
```
Comment (JustinPinner) [inline facia-press/app/frontpress/jsonModels.scala]: Maybe hasSlideshowImages or similar? Name suggests this is the actual images.

File: common/app/model/FaciaDisplayElement.scala [hunk 4]
```diff
@@ -45,3 +47,10 @@ case class CrosswordSvg(id: String) extends FaciaDisplayElement {
 
   def imageUrl = s"${Configuration.ajax.url}/$id.svg"
 }
+
+object InlineSlideshow {
+  def fromTrail(trail: Trail): InlineSlideshow =
+    InlineSlideshow(trail.trailSlideshow(5, 3))
```
Comment (JustinPinner) [inline common/app/model/FaciaDisplayElement.scala]: Discussed quickly with Jon, but maybe we could pass named aspect ratio vars instead of 5 and 3?

**Comment by JustinPinner** [comment]
Other than the questions I've asked, and the fact that some of the JS is a bit opaque to me, I think this generally looks ok. A second opinion would also be a good idea though!

**Comment by stephanfowler** [comment]
:+1: js looks good

File: static/src/javascripts/projects/common/modules/experiments/tests/facia-slideshow.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+], function (
+    detect
+) {
+    return function () {
+        this.id = 'FaciaSlideshow';
+        this.start = '2015-04-21';
+        this.expiry = '2015-05-21';
+        this.author = 'John Duffell';
+        this.description = 'Testing to see if slideshows affect page views per visit.';
+        this.audience = 0.99;
```
Comment (phamann) [inline static/src/javascripts/projects/common/modules/experiments/tests/facia-slideshow.js]: Why `0.99`????

**Pull request merged**